packages/casl-ability/spec/ability.spec.js

```diff
@@ -460,6 +460,43 @@ describe('Ability', () => {
     })
   })
 
+  describe('`manage` action', ()=>{
+    it('allows `all` action', () => {
+      ability = AbilityBuilder.define((can, cannot) => {
+        can('manage', 'all')
+      })
+
+      expect(ability).to.allow('read', 'post')
+    })
+
+    it('cannot is still honoured', () => {
+      ability = AbilityBuilder.define((can, cannot) => {
+        cannot('read', 'post')
+        can('manage', 'all')
+      })
+
+      expect(ability).not.to.allow('read', 'post')
+    })
+
+    it('`managcannote` can be used with cannot', () => {
+      ability = AbilityBuilder.define((can, cannot) => {
+        can('read', 'post')
+        cannot('manage', 'all')
+      })
+
+      expect(ability).to.allow('read', 'post')
```

Comment:
I'm concerned a bit that `cannot('manage', 'all')` is the last action which should override `can('read', 'post')` but it doesn't. I know that this is due to how `all` alias works but `manage` will work the same way and I think this is wrong behavior.

I need some time to think about this

Reply:
I dont quite get what you mean? At the moment rules are order specific.

```
can("read", "post")
cannot("read", "post")
```

You would expect it to return true, for me its the same with the alias

Reply:
From what you wrote I expect it to return false. 

In the test you say that user can read post and the next rule says it cannot do anything, so at logic level this doesn’t work :)

Try to change order of rules in the test, you will always receive the same result. This is incorrect behavior

Reply:
It depends on what your logic is ;)
If you say take the first rule that matches, then its fine.
If you say take the last rule that matches, then it's not.

So, are you saying that we need to:

- Check all rules that do not use an alias.
- If none found, check all rules that use an alias

In master, these two tests both pass, even though the rules are switched.

```
  it('`manage` can be used with cannot', () => {
      ability = AbilityBuilder.define((can, cannot) => {
       
        can('read', 'post')
        cannot('read', 'all')
      })

      expect(ability).to.allow('read', 'post')
      expect(ability).not.to.allow('delete', 'post')
    })
```

```
  it('`manage` can be used with cannot', () => {
      ability = AbilityBuilder.define((can, cannot) => {
       
        cannot('read', 'all')
        can('read', 'post')
       
      })

      expect(ability).to.allow('read', 'post')
      expect(ability).not.to.allow('delete', 'post')
    })
```

Reply:
Correct. 

 CASL takes the last rule which matches that’s why it wrong.the last rule overuses preceding one: check “combining rules” in https://stalniy.github.io/casl/abilities/2017/07/20/define-abilities.html#combining-rules

This is a mistake which I did in the initial commits but realized just now :)

Reply:
So the logic should be:

Get all the rules that are applicable.
Reverse them.
Take the first one that matches.

Reply:
Can you confirm the above before I start on it?

Reply:
Sorry was quite busy during the last 2 weeks, I plan to finalize & release this during this weekend. 

The main issue currently is that with such logic buildIndex function becomes more complicated. I started playing with the logic some time ago: https://jsbin.com/guvuxujeku/5/edit?js,console